# Churn Prediction Model

In [ ]:
!pip list

In [1]:
#importing Libraries
import pandas as pd
from sklearn import metrics
from sklearn.model_selection import train_test_split
from sklearn.metrics import recall_score
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.tree import DecisionTreeClassifier
from imblearn.combine import SMOTEENN
print("Libraries Imported")

Libraries Imported


#### Reading DataFrame 

In [2]:
df = pd.read_csv('telecom_churn.csv')
print('DataFrame Created')

DataFrame Created


In [3]:
df.head()

SeniorCitizen  MonthlyCharges  TotalCharges  Churn  gender_Female  \
0              0           29.85         29.85      0              1   
1              0           56.95       1889.50      0              0   
2              0           53.85        108.15      1              0   
3              0           42.30       1840.75      0              0   
4              0           70.70        151.65      1              1   

   gender_Male  Partner_No  Partner_Yes  Dependents_No  Dependents_Yes  ...  \
0            0           0            1              1               0  ...   
1            1           1            0              1               0  ...   
2            1           1            0              1               0  ...   
3            1           1            0              1               0  ...   
4            0           1            0              1               0  ...   

   PaymentMethod_Bank transfer (automatic)  \
0                                        0   
1                                        0   
2                                        0   
3                                        1   
4                                        0   

   PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
0                                      0                               1   
1                                      0                               0   
2                                      0                               0   
3                                      0                               0   
4                                      0                               1   

   PaymentMethod_Mailed check  tenure_group_1 - 12  tenure_group_13 - 24  \
0                           0                    1                     0   
1                           1                    0                     0   
2                           1                    1                     0   
3                           0                    0                     0   
4                           0                    1                     0   

   tenure_group_25 - 36  tenure_group_37 - 48  tenure_group_49 - 60  \
0                     0                     0                     0   
1                     1                     0                     0   
2                     0                     0                     0   
3                     0                     1                     0   
4                     0                     0                     0   

   tenure_group_61 - 72  
0                     0  
1                     0  
2                     0  
3                     0  
4                     0  

[5 rows x 51 columns]

In [4]:
#deleting unnamed column
# df.drop('Unnamed: 0',axis=1)

### Creating X and Y dependent variables

In [5]:
#x variable
x = df.drop('Churn',axis=1)
x

SeniorCitizen  MonthlyCharges  TotalCharges  gender_Female  gender_Male  \
0                 0           29.85         29.85              1            0   
1                 0           56.95       1889.50              0            1   
2                 0           53.85        108.15              0            1   
3                 0           42.30       1840.75              0            1   
4                 0           70.70        151.65              1            0   
...             ...             ...           ...            ...          ...   
7027              0           84.80       1990.50              0            1   
7028              0          103.20       7362.90              1            0   
7029              0           29.60        346.45              1            0   
7030              1           74.40        306.60              0            1   
7031              0          105.65       6844.50              0            1   

      Partner_No  Partner_Yes  Dependents_No  Dependents_Yes  PhoneService_No  \
0              0            1              1               0                1   
1              1            0              1               0                0   
2              1            0              1               0                0   
3              1            0              1               0                1   
4              1            0              1               0                0   
...          ...          ...            ...             ...              ...   
7027           0            1              0               1                0   
7028           0            1              0               1                0   
7029           0            1              0               1                1   
7030           0            1              1               0                0   
7031           1            0              1               0                0   

      ...  PaymentMethod_Bank transfer (automatic)  \
0     ...                                        0   
1     ...                                        0   
2     ...                                        0   
3     ...                                        1   
4     ...                                        0   
...   ...                                      ...   
7027  ...                                        0   
7028  ...                                        0   
7029  ...                                        0   
7030  ...                                        0   
7031  ...                                        1   

      PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
0                                         0                               1   
1                                         0                               0   
2                                         0                               0   
3                                         0                               0   
4                                         0                               1   
...                                     ...                             ...   
7027                                      0                               0   
7028                                      1                               0   
7029                                      0                               1   
7030                                      0                               0   
7031                                      0                               0   

      PaymentMethod_Mailed check  tenure_group_1 - 12  tenure_group_13 - 24  \
0                              0                    1                     0   
1                              1                    0                     0   
2                              1                    1                     0   
3                              0                    0                     0   
4                              0                    1              

In [6]:
#y variable
y=df['Churn']
y

0       0
1       0
2       1
3       0
4       1
       ..
7027    0
7028    0
7029    0
7030    1
7031    0
Name: Churn, Length: 7032, dtype: int64

## Creating variables for Training and Testing of Data
## Train Test Split

In [7]:
#Train and Test
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2)
# .2 means if we have 1000 record then 800 for training and 200 for testing

## Decision Tree Classifier

In [8]:
#Decision Tree Model
dt_model  = DecisionTreeClassifier(criterion = "gini",random_state = 100,max_depth=6, min_samples_leaf=8)
print('Decision Tree Classifier ')

Decision Tree Classifier 


Fitting Data in model

In [9]:
dt_model.fit(x_train,y_train)

DecisionTreeClassifier(max_depth=6, min_samples_leaf=8, random_state=100)

Predicting the churn on the testing data x_test

In [10]:
#predict variable
y_predict = dt_model.predict(x_test)
print('Success')

Success


In [11]:
#checking y_predict
y_predict

array([1, 1, 0, ..., 0, 0, 0])

## Comparing the y_predict value to the actual y value

In [12]:
dt_model.score(x_test,y_test)

0.7803837953091685

Classification Model Report

In [13]:
print(classification_report(y_test, y_predict, labels=[0,1]))

              precision    recall  f1-score   support

           0       0.84      0.87      0.85      1041
           1       0.59      0.54      0.56       366

    accuracy                           0.78      1407
   macro avg       0.71      0.70      0.71      1407
weighted avg       0.77      0.78      0.78      1407



0 - Non Churners
1 - Churners
Model Accouracy is all about 78% which is not good because of imbalanced data.

Printing Confusion Matrix : 
True Positive (TP)        False Positive (FP)
Fasle Negative (FN)       True Negative (TN)

Accuracy = (TP+TN)/Total

In [14]:
#confusion Matrix
print(confusion_matrix(y_test,y_predict))

[[902 139]
 [170 196]]


In [15]:
#printing Accuracy Precentage
print((913+198)*100/(103+193+913+198))

78.96233120113718


## Conclusion of Decision Tree Model
Accuracy of Model is not good because of Imbalanced Data

Above Model's accuracy is not so good because of imbalancesd data
so we are using SMOTEENN for better results

## Model Using SMOTEENN

In [16]:
from imblearn.combine import SMOTEENN

In [17]:
sm = SMOTEENN()
X_resampled, y_resampled = sm.fit_resample(x,y)

Creating new variables for model

In [18]:
#SMOTEENN Model variables
xr_train,xr_test,yr_train,yr_test=train_test_split(X_resampled, y_resampled,test_size=0.2)


Creating SMOTEENN Model

In [19]:
#creating Model
model_dt_smote=DecisionTreeClassifier(criterion = "gini",random_state = 100,max_depth=6, min_samples_leaf=8)
print('Model Created')

Model Created


Fitting data in Model

In [20]:
#fitting data in model
model_dt_smote.fit(xr_train,yr_train)

DecisionTreeClassifier(max_depth=6, min_samples_leaf=8, random_state=100)

In [21]:
#predicting the variable
yr_predict = model_dt_smote.predict(xr_test)

In [22]:
#checking predicted variable values
yr_predict

array([0, 0, 1, ..., 0, 0, 1])

### Checking Score of this Model

In [23]:
#Model Score
model_score_r = model_dt_smote.score(xr_test, yr_test)
model_score_r 

0.9343434343434344

##### Accuracy is 92.7%

#### Model report

In [24]:
print(metrics.classification_report(yr_test, yr_predict))

              precision    recall  f1-score   support

           0       0.94      0.91      0.93       545
           1       0.93      0.95      0.94       643

    accuracy                           0.93      1188
   macro avg       0.94      0.93      0.93      1188
weighted avg       0.93      0.93      0.93      1188



Confusion Matrix

In [25]:
#confusion Matrix
print(metrics.confusion_matrix(yr_test, yr_predict))

[[498  47]
 [ 31 612]]


In [26]:
#printing Accuracy Precentage
print("Accuracy : ",(498+586)*100/(498+586+62+23))


Accuracy :  92.72882805816937


## Random Forest Classifier

In [27]:
from sklearn.ensemble import RandomForestClassifier

#### Creating Model

In [28]:
#Random Forest Model
model_rf = RandomForestClassifier(n_estimators=100, criterion='gini', random_state = 100,max_depth=6, min_samples_leaf=8)
print("Random Forest Model Created")

Random Forest Model Created


Fitting Data in Model

In [29]:
#data Fitting
model_rf.fit(x_train,y_train)

RandomForestClassifier(max_depth=6, min_samples_leaf=8, random_state=100)

In [30]:
#predicting values
y_pred = model_rf.predict(x_test)
y_pred 

array([1, 0, 0, ..., 0, 0, 0])

Model Score

In [31]:
model_rf.score(x_test,y_test)

0.7924662402274343

#### Accuracy is about 80%

In [32]:
#Classification report
print(classification_report(y_test, y_pred, labels=[0,1]))

              precision    recall  f1-score   support

           0       0.82      0.92      0.87      1041
           1       0.65      0.44      0.52       366

    accuracy                           0.79      1407
   macro avg       0.74      0.68      0.70      1407
weighted avg       0.78      0.79      0.78      1407



#confusion Matrix
print(metrics.confusion_matrix(x_test, y_pred))

In [33]:
#confusion Matrix
print(metrics.confusion_matrix(y_test,y_pred))

[[955  86]
 [206 160]]


In [34]:
#printing Accuracy Precentage
print("Accuracy : ",(958+181)*100/(958+181+84+184))

Accuracy :  80.95238095238095


###### Model Accuracy is 80.95%

### Creating new Model using SMOTEENN

In [35]:
#creating new variables
sm = SMOTEENN()
X_resampled1, y_resampled1 = sm.fit_resample(x,y)
print('Variable Created')

Variable Created


In [36]:
#Model Training
xr_train1,xr_test1,yr_train1,yr_test1 = train_test_split(X_resampled1, y_resampled1,test_size=0.2)

#### Creating Model 

In [37]:
model_rf_smote = RandomForestClassifier(n_estimators=100, criterion='gini', random_state = 100,max_depth=6, min_samples_leaf=8)
print("Random Forest SMOTEENN Model created")

Random Forest SMOTEENN Model created


In [38]:
#Fitting data in Model 
model_rf_smote.fit(xr_train1,yr_train1)

RandomForestClassifier(max_depth=6, min_samples_leaf=8, random_state=100)

Predict variable

In [39]:
#predict variable
yr_predict1 = model_rf_smote.predict(xr_test1)
yr_predict1

array([0, 0, 1, ..., 1, 1, 1])

#### Model Score

In [40]:
#Model Score
model_score_r1 = model_rf_smote.score(xr_test1, yr_test1)
model_score_r1 

0.9362786745964317

Accuracy : 93.7%

#### Model Report

In [41]:
print(metrics.classification_report(yr_test1, yr_predict1))

              precision    recall  f1-score   support

           0       0.94      0.92      0.93       542
           1       0.93      0.95      0.94       635

    accuracy                           0.94      1177
   macro avg       0.94      0.93      0.94      1177
weighted avg       0.94      0.94      0.94      1177



Conclusion : The Model's accuracy is better than other models
Precision of Churn     : 96%
Precision of Not Churn : 92%

In [42]:
#confusion Matrix
print(metrics.confusion_matrix(yr_test1, yr_predict1))

[[496  46]
 [ 29 606]]


In [118]:
#printing Accuracy Precentage
print("Accuracy : ",(473+623)*100/(473+623+51+22))

Accuracy :  93.75534644995723


## Saving Our Model using Pickle

In [43]:
import pickle

In [44]:
filename = "randomForestModel.sav"

In [45]:
pickle.dump(model_rf_smote,open(filename,'wb'))
print('Model Dumped')

Model Dumped


Checking Our Model

In [46]:
#loading Model
load_model = pickle.load(open(filename,'rb'))
print("Model Loaded")

Model Loaded


In [47]:
load_model.score(xr_test,yr_test)

0.9393939393939394